services/api/service.go

```diff
@@ -635,6 +751,37 @@ func (api *RelayAPI) updateProposerDuties(headSlot uint64) {
 	api.log.Infof("proposer duties updated: %s", strings.Join(_duties, ", "))
 }
 
+func (api *RelayAPI) prepareBuildersForSlot(headSlot uint64) {
+	// Wait until there are no optimistic blocks being processed. Then we can
+	// safely update the slot.
+	api.optimisticBlocksWG.Wait()
+	api.optimisticSlot.Store(headSlot + 1)
+
+	builders, err := api.db.GetBlockBuilders()
+	if err != nil {
+		api.log.WithError(err).Error("unable to read block builders from db, not updating builder cache")
+		return
+	}
+	for _, v := range builders {
+		entry := &blockBuilderCacheEntry{ //nolint:exhaustruct
+			status: common.BuilderStatus{
+				IsHighPrio:    v.IsHighPrio,
+				IsBlacklisted: v.IsBlacklisted,
+				IsOptimistic:  v.IsOptimistic,
+			},
+		}
+		// Try to parse builder collateral string to big int.
+		builderCollateral, ok := big.NewInt(0).SetString(v.Collateral, 10)
+		if !ok {
+			api.log.WithError(err).Errorf("could not parse builder collateral string %s", v.Collateral)
+			entry.collateral = big.NewInt(0)
+		} else {
+			entry.collateral = builderCollateral
+		}
+		api.blockBuildersCache[v.BuilderPubkey] = entry
```

Comment:
do we need a lock for updating the `blockBuildersCache`, or can we ignore any concurrency issues because this function will not be called concurrently?

Reply:
also, should we build a new list and replace `blockBuildersCache`, rather than just updating the individual entries? if a builder would be removed from the database (unlikely) then it would still be persisted here.